IOBalancerMemoryLeakTest.testMemoryLeak

```
java.net.SocketException: Unexpected end of file from server
	at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:776)
	at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:633)
	at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:773)
	at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:633)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1324)
	at com.hazelcast.internal.ascii.HTTPCommunicator.doGet(HTTPCommunicator.java:140)
	at com.hazelcast.internal.ascii.HTTPCommunicator.getClusterInfo(HTTPCommunicator.java:81)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest.testMemoryLeak(IOBalancerMemoryLeakTest.java:52)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/745/testReport/junit/com.hazelcast.nio.tcp.nonblocking.iobalancer/IOBalancerMemoryLeakTest/testMemoryLeak/

This happened when I executed all tests locally to measure code coverage (which also picks up nightly tests):
```
java.lang.AssertionError: 
Expected :0
Actual   :4
 <Click to see difference>


	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest$1.run(IOBalancerMemoryLeakTest.java:65)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:835)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:849)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest.testMemoryLeak(IOBalancerMemoryLeakTest.java:59)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at com.hazelcast.test.AbstractHazelcastClassRunner$ThreadDumpAwareRunAfters.evaluate(AbstractHazelcastClassRunner.java:132)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at com.hazelcast.test.HazelcastSerialClassRunner.runChild(HazelcastSerialClassRunner.java:50)
	at com.hazelcast.test.HazelcastSerialClassRunner.runChild(HazelcastSerialClassRunner.java:29)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at com.hazelcast.test.AbstractHazelcastClassRunner$1.evaluate(AbstractHazelcastClassRunner.java:184)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
```

```
java.lang.AssertionError: expected:<0> but was:<3>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest$1.run(IOBalancerMemoryLeakTest.java:65)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:850)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:864)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest.testMemoryLeak(IOBalancerMemoryLeakTest.java:60)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/836/testReport/junit/com.hazelcast.nio.tcp.nonblocking.iobalancer/IOBalancerMemoryLeakTest/testMemoryLeak/

```
java.lang.AssertionError: expected:<0> but was:<3>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest$1.run(IOBalancerMemoryLeakTest.java:65)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:851)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:865)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest.testMemoryLeak(IOBalancerMemoryLeakTest.java:60)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/855/testReport/junit/com.hazelcast.nio.tcp.nonblocking.iobalancer/IOBalancerMemoryLeakTest/testMemoryLeak/

```
java.lang.AssertionError: expected:<0> but was:<2>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest$1.run(IOBalancerMemoryLeakTest.java:65)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:851)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:865)
	at com.hazelcast.nio.tcp.nonblocking.iobalancer.IOBalancerMemoryLeakTest.testMemoryLeak(IOBalancerMemoryLeakTest.java:60)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/860/testReport/junit/com.hazelcast.nio.tcp.nonblocking.iobalancer/IOBalancerMemoryLeakTest/testMemoryLeak/

--- a/hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/IOBalancer.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/IOBalancer.java
@@ -103,27 +103,13 @@ public class IOBalancer {
     }
 
     public void connectionAdded(TcpIpConnection connection) {
-        NonBlockingSocketReader socketReader = (NonBlockingSocketReader) connection.getSocketReader();
-        NonBlockingSocketWriter socketWriter = (NonBlockingSocketWriter) connection.getSocketWriter();
-
-        if (logger.isFinestEnabled()) {
-            logger.finest("Added handlers for: " + connection);
-        }
-
-        inLoadTracker.addHandler(socketReader);
-        outLoadTracker.addHandler(socketWriter);
+        inLoadTracker.notifyConnectionAdded(connection);
+        outLoadTracker.notifyConnectionAdded(connection);
     }
 
     public void connectionRemoved(TcpIpConnection connection) {
-        NonBlockingSocketReader socketReader = (NonBlockingSocketReader) connection.getSocketReader();
-        NonBlockingSocketWriter socketWriter = (NonBlockingSocketWriter) connection.getSocketWriter();
-
-        if (logger.isFinestEnabled()) {
-            logger.finest("Removing handlers from: " + connection);
-        }
-
-        inLoadTracker.removeHandler(socketReader);
-        outLoadTracker.removeHandler(socketWriter);
+        inLoadTracker.notifyConnectionRemoved(connection);
+        outLoadTracker.notifyConnectionRemoved(connection);
     }
 
     public void start() {
@@ -211,4 +197,5 @@ public class IOBalancer {
     public void signalMigrationComplete() {
         migrationCompletedCount.inc();
     }
+
 }

--- a/hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/LoadTracker.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/LoadTracker.java
@@ -17,15 +17,19 @@
 package com.hazelcast.nio.tcp.nonblocking.iobalancer;
 
 import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.tcp.TcpIpConnection;
 import com.hazelcast.nio.tcp.nonblocking.MigratableHandler;
 import com.hazelcast.nio.tcp.nonblocking.NonBlockingIOThread;
+import com.hazelcast.nio.tcp.nonblocking.NonBlockingSocketReader;
+import com.hazelcast.nio.tcp.nonblocking.NonBlockingSocketWriter;
 import com.hazelcast.util.ItemCounter;
-
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.Iterator;
 import java.util.Map;
+import java.util.Queue;
 import java.util.Set;
-import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.LinkedBlockingQueue;
 
 import static com.hazelcast.util.StringUtil.LINE_SEPARATOR;
 
@@ -52,9 +56,11 @@ class LoadTracker {
     private final ItemCounter<MigratableHandler> handlerEventsCounter = new ItemCounter<MigratableHandler>();
 
     //contains all known handlers
-    private final Set<MigratableHandler> handlers = new CopyOnWriteArraySet<MigratableHandler>();
+    private final Set<MigratableHandler> handlers = new HashSet<MigratableHandler>();
 
     private final LoadImbalance imbalance;
+
+    private final Queue<Runnable> tasks = new LinkedBlockingQueue<Runnable>();
 
     LoadTracker(NonBlockingIOThread[] ioThreads, ILogger logger) {
         this.logger = logger;
@@ -76,6 +82,7 @@ class LoadTracker {
      * @return recalculated imbalance
      */
     LoadImbalance updateImbalance() {
+        handleAddedOrRemovedConnections();
         clearWorkingImbalance();
         updateNewWorkingImbalance();
         updateNewFinalImbalance();
@@ -83,9 +90,28 @@ class LoadTracker {
         return imbalance;
     }
 
+    private void handleAddedOrRemovedConnections() {
+        Iterator<Runnable> iterator = tasks.iterator();
+        while (iterator.hasNext()) {
+            Runnable task = iterator.next();
+            task.run();
+            iterator.remove();
+        }
+    }
+
     // just for testing
     Set<MigratableHandler> getHandlers() {
         return handlers;
+    }
+
+    // just for testing
+    ItemCounter<MigratableHandler> getLastEventCounter() {
+        return lastEventCounter;
+    }
+
+    // just for testing
+    ItemCounter<MigratableHandler> getHandlerEventsCounter() {
+        return handlerEventsCounter;
     }
 
     private void updateNewFinalImbalance() {
@@ -111,6 +137,17 @@ class LoadTracker {
             }
         }
     }
+
+    public void notifyConnectionAdded(TcpIpConnection connection) {
+        ConnectionAddedTask connectionAddedTask = new ConnectionAddedTask(connection);
+        tasks.offer(connectionAddedTask);
+    }
+
+    public void notifyConnectionRemoved(TcpIpConnection connection) {
+        ConnectionRemovedTask connectionRemovedTask = new ConnectionRemovedTask(connection);
+        tasks.offer(connectionRemovedTask);
+    }
+
 
     private void updateNewWorkingImbalance() {
         for (MigratableHandler handler : handlers) {
@@ -147,6 +184,8 @@ class LoadTracker {
 
     void removeHandler(MigratableHandler handler) {
         handlers.remove(handler);
+        handlerEventsCounter.remove(handler);
+        lastEventCounter.remove(handler);
     }
 
     private void printDebugTable() {
@@ -217,4 +256,49 @@ class LoadTracker {
         }
         sb.append(LINE_SEPARATOR);
     }
+
+    class ConnectionRemovedTask implements Runnable {
+
+        private TcpIpConnection connection;
+
+        public ConnectionRemovedTask(TcpIpConnection connection) {
+            this.connection = connection;
+        }
+
+        @Override
+        public void run() {
+            NonBlockingSocketReader socketReader = (NonBlockingSocketReader) connection.getSocketReader();
+            NonBlockingSocketWriter socketWriter = (NonBlockingSocketWriter) connection.getSocketWriter();
+
+            if (logger.isFinestEnabled()) {
+                logger.finest("Removing handlers from: " + connection);
+            }
+
+            removeHandler(socketReader);
+            removeHandler(socketWriter);
+        }
+    }
+
+    class ConnectionAddedTask implements Runnable {
+
+        private TcpIpConnection connection;
+
+        public ConnectionAddedTask(TcpIpConnection connection) {
+            this.connection = connection;
+        }
+
+        @Override
+        public void run() {
+            NonBlockingSocketReader socketReader = (NonBlockingSocketReader) connection.getSocketReader();
+            NonBlockingSocketWriter socketWriter = (NonBlockingSocketWriter) connection.getSocketWriter();
+
+            if (logger.isFinestEnabled()) {
+                logger.finest("Added handlers for: " + connection);
+            }
+
+            addHandler(socketReader);
+            addHandler(socketWriter);
+        }
+    }
+
 }

--- a/hazelcast/src/main/java/com/hazelcast/util/ItemCounter.java
+++ b/hazelcast/src/main/java/com/hazelcast/util/ItemCounter.java
@@ -113,6 +113,10 @@ public final class ItemCounter<T> {
         return oldValue;
     }
 
+    public void remove(T item) {
+        map.remove(item);
+    }
+
     @Override
     public boolean equals(Object o) {
         if (this == o) {

--- a/hazelcast/src/test/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/IOBalancerMemoryLeakTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/nio/tcp/nonblocking/iobalancer/IOBalancerMemoryLeakTest.java
@@ -22,12 +22,15 @@ import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.instance.GroupProperty;
 import com.hazelcast.instance.HazelcastInstanceManager;
 import com.hazelcast.internal.ascii.HTTPCommunicator;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.Protocols;
 import com.hazelcast.nio.tcp.TcpIpConnectionManager;
 import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.NightlyTest;
 import java.io.IOException;
+import java.net.Socket;
 import org.junit.After;
 import org.junit.Assert;
 import org.junit.Before;
@@ -46,10 +49,12 @@ public class IOBalancerMemoryLeakTest extends HazelcastTestSupport {
     }
 
     @Test
-    public void testMemoryLeak() throws IOException {
+    public void testMemoryLeak_with_RestConnections() throws IOException {
         Config config = new Config();
         config.getGroupConfig().setName(randomName());
         config.setProperty(GroupProperty.REST_ENABLED, "true");
+        config.setProperty(GroupProperty.IO_BALANCER_INTERVAL_SECONDS, "1");
+
         HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
         HTTPCommunicator communicator = new HTTPCommunicator(instance);
         TcpIpConnectionManager connectionManager = (TcpIpConnectionManager) getConnectionManager(instance);
@@ -67,5 +72,65 @@ public class IOBalancerMemoryLeakTest extends HazelcastTestSupport {
             }
         });
     }
+
+
+    @Test
+    public void testMemoryLeak_with_SocketConnections() throws IOException {
+        Config config = new Config();
+        config.getGroupConfig().setName(randomName());
+        config.setProperty(GroupProperty.IO_BALANCER_INTERVAL_SECONDS, "1");
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+        final Address address = instance.getCluster().getLocalMember().getAddress();
+        int threadCount = 10;
+        final int connectionCountPerThread = 100;
+
+        Runnable runnable = new Runnable() {
+            public void run() {
+                for (int i = 0; i < connectionCountPerThread; i++) {
+                    Socket socket = null;
+                    try {
+                        socket = new Socket(address.getHost(), address.getPort());
+                        socket.getOutputStream().write(Protocols.CLUSTER.getBytes());
+                        sleepMillis(1000);
+                        socket.close();
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                }
+            }
+        };
+
+        Thread[] threads = new Thread[threadCount];
+        for (int i = 0; i < threadCount; i++) {
+            threads[i] = new Thread(runnable);
+            threads[i].start();
+        }
+
+        assertJoinable(threads);
+
+
+        TcpIpConnectionManager connectionManager = (TcpIpConnectionManager) getConnectionManager(instance);
+        final IOBalancer ioBalancer = connectionManager.getIoBalancer();
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                LoadTracker inLoadTracker = ioBalancer.getInLoadTracker();
+                LoadTracker outLoadTracker = ioBalancer.getOutLoadTracker();
+                int inHandlerSize = inLoadTracker.getHandlers().size();
+                int outHandlerSize = outLoadTracker.getHandlers().size();
+                int inHandlerEventsCount = inLoadTracker.getHandlerEventsCounter().keySet().size();
+                int outHandlerEventsCount = outLoadTracker.getHandlerEventsCounter().keySet().size();
+                int inLastEventsCount = inLoadTracker.getLastEventCounter().keySet().size();
+                int outLastEventsCount = outLoadTracker.getLastEventCounter().keySet().size();
+                Assert.assertEquals(0, inHandlerSize);
+                Assert.assertEquals(0, outHandlerSize);
+                Assert.assertEquals(0, inHandlerEventsCount);
+                Assert.assertEquals(0, outHandlerEventsCount);
+                Assert.assertEquals(0, inLastEventsCount);
+                Assert.assertEquals(0, outLastEventsCount);
+            }
+        });
+    }
+
 
 }

